File: cmd/bx/ingress.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

// Ingress commands (plans/ingress.md): publishing tiles.
//
//	bx expose <tile> <slot>=<source> [--host H | --zone '*.Z'] [--listen :P] [--add]
//	bx unexpose <tile> <slot> [--host H | --zone '*.Z' | --listen :P]
//	bx ingress            published endpoints, routes, listener status
//	bx ingress routes     just the host → tile routes

func cmdExpose(args []string) error {
	var pos []string
	body := map[string]any{}
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--add": // one more route on the endpoint, beside the ones it has (D79)
			body["add"] = true
		case "--host", "--zone", "--listen", "--tcp", "--udp":
			if i+1 >= len(args) {
				return fmt.Errorf("%s needs a value", args[i])
			}
			key := strings.TrimPrefix(args[i], "--")
			if key == "tcp" || key == "udp" {
				key = "listen" // --tcp/--udp are listen-addr aliases (proto comes from the manifest)
			}
			body[key] = args[i+1]
			i++
		default:
			pos = append(pos, args[i])
		}
	}
	if len(pos) != 2 || !strings.Contains(pos[1], "=") {
		return fmt.Errorf(`usage: bx expose <tile> <slot>=<source> [--host <name> | --zone '*.<suffix>'] [--listen :<port>] [--add]
  sources: "runtime" (xbind's listener / a host port) or an ingress terminator tile
  --add adds a route beside the endpoint's others (more hostnames, more host
  ports); without it the endpoint's routes are replaced by this one
  examples:
    bx expose apps/blog web=apps/traefik --host blog.example.com
    bx expose apps/blog web=apps/traefik --host shop.example.com --add
    bx expose apps/cms  web=apps/traefik --zone '*.sites.example.com'
    bx expose apps/blog web=runtime --host blog.example.com
    bx expose apps/game game=runtime --listen :2456`)
	}
	slot, source, _ := strings.Cut(pos[1], "=")
	body["component"], body["slot"], body["provider"] = pos[0], slot, source
	if err := apiJSON("POST", "/api/xbin/bindings", body, nil); err != nil {
		return err
	}
	fmt.Println("published — check `bx ingress`")
	return nil
}

func cmdUnexpose(args []string) error {
	var pos []string
	body := map[string]string{}
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--host", "--zone", "--listen":
			if i+1 >= len(args) {
				return fmt.Errorf("%s needs a value", args[i])
			}
			body[strings.TrimPrefix(args[i], "--")] = args[i+1]
			i++
		default:
			pos = append(pos, args[i])
		}
	}
	if len(pos) != 2 {
		return fmt.Errorf("usage: bx unexpose <tile> <slot> [--host <name> | --zone '*.<suffix>' | --listen :<port>]\n  with a route: remove just that one; without: every route of the endpoint")
	}
	body["component"], body["slot"] = pos[0], pos[1]
	if err := apiJSON("DELETE", "/api/xbin/bindings", body, nil); err != nil {
		return err
	}
	if len(body) > 2 {
		fmt.Println("route removed")
	} else {
		fmt.Println("unpublished")
	}
	return nil
}

type exposeRouteOut struct {
	Source string `json:"source"`
	Host   string `json:"host"`
	Zone   string `json:"zone"`
	Listen string `json:"listen"`
}

type ingressOut struct {
	Exposes []struct {
		Component string           `json:"component"`
		Slot      string           `json:"slot"`
		Kind      string           `json:"kind"`
		Paths     []string         `json:"paths"`
		Proto     string           `json:"proto"`
		Port      int              `json:"port"`
		Source    string           `json:"source"`
		Host      string           `json:"host"`
		Zone      string           `json:"zone"`
		Listen    string           `json:"listen"`
		Blocked   string           `json:"blocked"`
		Routes    []exposeRouteOut `json:"routes"` // every route (an older xbind: absent — the scalars hold its one)
	} `json:"exposes"`
	Routes []struct {
		Host      string `json:"host"`
		Component string `json:"component"`
		Slot      string `json:"slot"`
		Source    string `json:"source"`
		Zone      string `json:"zone"`
	} `json:"routes"`
	Streams []struct {
		Component string `json:"component"`
		Slot      string `json:"slot"`
		Proto     string `json:"proto"`
		Listen    string `json:"listen"`
		Port      int    `json:"port"`
		Error     string `json:"error"`
		Active    int    `json:"active"`
	} `json:"streams"`
	Forwards []struct {
		Source string `json:"source"`
		Error  string `json:"error"`
	} `json:"forwards"`
	HTTPListener struct {
		Listen string `json:"listen"`
		TLS    bool   `json:"tls"`
	} `json:"httpListener"`
	IngressHosts map[string][]string `json:"ingressHosts"`
}

func cmdIngress(args []string) error {
	var out ingressOut
	if err := apiJSON("GET", "/api/xbin/ingress", nil, &out); err != nil {
		return err
	}
	routesOnly := len(args) > 0 && args[0] == "routes"

	if !routesOnly {
		fmt.Println("published endpoints:")
		if len(out.Exposes) == 0 {
			fmt.Println("  (none — tiles declare \"exposes\" in xbin.json; see /docs/ingress.md)")
		}
		for _, e := range out.Exposes {
			routes := e.Routes
			if len(routes) == 0 && e.Source != "" {
				routes = []exposeRouteOut{{e.Source, e.Host, e.Zone, e.Listen}}
			}
			var states []string // one per route: an endpoint takes many (D79)
			for _, r := range routes {
				if e.Kind == "http" {
					where := r.Host
					if r.Zone != "" {
						where = r.Zone + " (delegated zone)"
					}
					states = append(states, fmt.Sprintf("→ %s  %s", r.Source, where))
					continue
				}
				listen := r.Listen
				if listen == "" {
					listen = fmt.Sprintf(":%d", e.Port)
				}
				states = append(states, fmt.Sprintf("→ host %s/%s → :%d", listen, e.Proto, e.Port))
			}
			switch {
			case e.Blocked != "":
				states = []string{"BLOCKED: " + e.Blocked}
			case len(states) == 0:
				states = []string{"(unbound — not reachable)"}
			}
			detail := ""
			if e.Kind == "http" && len(e.Paths) > 0 {
				detail = "  public: " + strings.Join(e.Paths, " ")
			}
			fmt.Printf("  %-24s %-10s %-6s %s%s\n", e.Component, e.Slot, e.Kind, states[0], detail)
			for _, s := range states[1:] {
				fmt.Printf("  %-24s %-10s %-6s %s\n", "", "", "", s)
			}
		}
	}

	fmt.Println("routes:")
	if len(out.Routes) == 0 {
		fmt.Println("  (none)")
	}
	for _, r := range out.Routes {
		via := r.Source
		if r.Zone != "" {
			via += " (zone " + r.Zone + ")"
		}
		fmt.Printf("  %-32s → %s.%s  via %s\n", r.Host, r.Component, r.Slot, via)
	}
	if routesOnly {
		return nil
	}

	if len(out.Streams) > 0 {
		fmt.Println("stream listeners:")
		for _, s := range out.Streams {
			status := fmt.Sprintf("%d active", s.Active)
			if s.Error != "" {
				status = "ERROR: " + s.Error
			}
			fmt.Printf("  %-20s %s/%s → %s.%s:%d  %s\n", s.Listen, s.Proto, "host", s.Component, s.Slot, s.Port, status)
		}
	}
	if len(out.Forwards) > 0 {
		fmt.Println("terminator forward doors:")
		for _, f := range out.Forwards {
			status := "up"
			if f.Error != "" {
				status = "ERROR: " + f.Error
			}
			fmt.Printf("  %-24s %s\n", f.Source, status)
		}
	}
	if out.HTTPListener.Listen != "" {
		tls := "no TLS (front it or use the traefik tile)"
		if out.HTTPListener.TLS {
			tls = "TLS on"
		}
		fmt.Printf("builtin HTTP listener: %s (%s)\n", out.HTTPListener.Listen, tls)
	} else {
		fmt.Println("builtin HTTP listener: off (start xbind with --ingress-listen)")
	}
	if len(out.IngressHosts) > 0 {
		fmt.Println("zone registrations:")
		comps := make([]string, 0, len(out.IngressHosts))
		for c := range out.IngressHosts {
			comps = append(comps, c)
		}
		sort.Strings(comps)
		for _, c := range comps {
			fmt.Printf("  %-24s %s\n", c, strings.Join(out.IngressHosts[c], " "))
		}
	}
	return nil
}

File: workspace-template/tiles/admin/tabs/ingress.js
/**
 * <bx-admin-ingress view="expose|endpoints"> — the admin console's ingress
 * group: publish a tile's declared exposes to an ingress source (host / zone
 * / host port), and the live public routing table (routes, stream
 * listeners, terminator doors). Loads /ingress itself; reports through
 * bx-admin-err / bx-admin-tab.
 */
import { LitElement, html, nothing } from 'lit';
import { xbinApi as api } from '/vendor/bx-kit.js';
import { base } from '../admin-css.js';
import { WithFilter, WithRouter } from '../shared.js';

export class BxAdminIngress extends WithRouter(WithFilter(LitElement)) {
  static properties = {
    view: { type: String },     // expose | endpoints
    _ingress: { state: true },  // /ingress {exposes, routes, streams, forwards, httpListener, terminators}
    _ingEdit: { state: true },  // per-endpoint add-a-route edits (comp\x00slot → {source, host, zone, listen})
    _q: { state: true },
    _cats: { state: true },
    _err: { state: true },
  };
  static styles = [base];

  constructor() {
    super();
    this._q = ''; this._cats = new Set();
  }
  connectedCallback() {
    super.connectedCallback();
    this.load();
  }
  refresh() { return this.load(); }
  async load() {
    try { this._ingress = await api('/ingress'); this._ok(); } catch (e) { this._fail(e); }
  }

  render() { return this.view === 'endpoints' ? this._ingressEndpointsView() : this._ingressExposeView(); }

  _ingKey(comp, slot) { return comp + '\x00' + slot; }
  // every route of an exposed endpoint — it takes many (D79: more hostnames,
  // more host ports); an older xbind sends only the first, in the scalars
  _ingRoutes(e) {
    if (Array.isArray(e.routes)) return e.routes;
    return e.source ? [{ source: e.source, host: e.host, zone: e.zone, listen: e.listen }] : [];
  }
  // the add-a-route editor's working state (per endpoint)
  _ingEditFor(e) { return this._ingEdit?.[this._ingKey(e.component, e.slot)] ?? { source: '', host: '', zone: '', listen: '' }; }
  _ingSetEdit(e, patch) {
    const k = this._ingKey(e.component, e.slot);
    this._ingEdit = { ...(this._ingEdit || {}), [k]: { ...this._ingEditFor(e), ...patch } };
  }
  async _ingPublish(e) {
    const ed = this._ingEditFor(e);
    if (!ed.source) return;
    // a bound endpoint gets one MORE route; an unbound one is published
    const body = { component: e.component, slot: e.slot, provider: ed.source, add: this._ingRoutes(e).length > 0 };
    if (e.kind === 'http') {
      // Exactly one of host/zone — send whichever is filled (server validates).
      if (ed.zone) body.zone = ed.zone; else body.host = ed.host;
    } else if (ed.listen) body.listen = ed.listen;
    try {
      await api('/bindings', { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(body) });
      const k = this._ingKey(e.component, e.slot);
      const { [k]: _, ...rest } = this._ingEdit || {}; this._ingEdit = rest;
      await this.load();
    } catch (err) { this._fail(err); }
  }
  // remove one route (the endpoint stays published while any is left)
  async _ingRemove(e, r) {
    const body = { component: e.component, slot: e.slot, provider: r.source };
    if (r.host) body.host = r.host;
    if (r.zone) body.zone = r.zone;
    if (r.listen) body.listen = r.listen;
    try {
      await api('/bindings', { method: 'DELETE', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(body) });
      await this.load();
    } catch (err) { this._fail(err); }
  }
  // ---- ingress → services / expose: publish tiles to the outside ----
  _ingressExposeView() {
    const d = this._ingress;
    if (!d) return html`<span class="muted">loading…</span>`;
    const streams = d.streams || [];
    const sources = (kind) => kind === 'http' ? ['runtime', ...(d.terminators || [])] : ['runtime'];
    const all = d.exposes || [];
    const exposes = all.filter((e) => this._match(e.component, e.slot, e.kind,
      ...this._ingRoutes(e).flatMap((r) => [r.source, r.host, r.zone, r.listen])));
    const routeText = (e, r) => e.kind === 'http' ? (r.zone ? `${r.zone} (zone)` : r.host) : `host ${r.listen || ':' + e.port} → :${e.port}`;
    const routeState = (e, r) => {
      if (e.kind === 'http') return html`<span class="st-healthy">public: ${r.zone || r.host}</span>`;
      const listen = r.listen || ':' + e.port;
      const st = streams.find((s) => s.component === e.component && s.slot === e.slot && s.listen === listen);
      return st?.error ? html`<span class="st-failed">⚠ ${st.error}</span>`
        : html`<span class="st-healthy">listening${st ? ` (${st.active} active)` : ''}</span>`;
    };
    return html`
      <p class="muted">Tiles declare <code>exposes</code> in their manifest; <b>binding a slot to an
        ingress source publishes it</b> to the outside — anonymous traffic, confined to the tile's
        declared public paths. Unbound = unreachable, exactly like interfaces. An endpoint takes
        <b>many routes</b>: more hostnames (through one terminator or several), more host ports —
        each hostname and port still belongs to exactly one endpoint.
        See <a href="/docs/ingress.md" target="_blank">docs/ingress.md</a>.</p>
      ${all.length === 0 ? html`<div class="muted">No tile declares <code>exposes</code> yet. Add an
        <span class="mono">exposes</span> block to a tile's <span class="mono">xbin.json</span>
        (http paths, or a tcp/udp port), or import the <b>Public HTTPS (Traefik)</b> tile.</div>` : html`
      ${this._filterBar('filter exposed endpoints…', null, exposes.length, all.length)}
      <table class="tbl">
        <tr><th>tile</th><th>endpoint</th><th>source</th><th>route</th><th></th><th>state</th></tr>
        ${exposes.map((e) => {
          const routes = this._ingRoutes(e);
          const ed = this._ingEditFor(e);
          const endpoint = e.kind === 'http'
            ? html`<span class="pill">http</span> <span class="muted">${(e.paths || []).join(' ')}</span>`
            : html`<span class="pill">${e.proto}:${e.port}</span>`;
          const head = (i) => i === 0 ? html`<td class="mono">${e.component}</td><td>${e.slot} ${endpoint}</td>` : html`<td></td><td></td>`;
          const routeEd = e.kind === 'http'
            ? html`<input class="mono ing-host" style="width:12em" placeholder="host (blog.example.com)"
                     .value=${ed.zone ? '' : ed.host} ?disabled=${!!ed.zone}
                     @input=${(ev) => this._ingSetEdit(e, { host: ev.target.value.trim(), zone: '' })}>
                   <input class="mono ing-zone" style="width:11em" placeholder="or zone (*.sites.…)"
                     .value=${ed.zone}
                     @input=${(ev) => this._ingSetEdit(e, { zone: ev.target.value.trim() })}>`
            : html`<input class="mono ing-listen" style="width:8em" placeholder=":${e.port} (host port)"
                     .value=${ed.listen}
                     @input=${(ev) => this._ingSetEdit(e, { listen: ev.target.value.trim() })}>`;
          const addState = e.blocked ? html`<span class="st-failed">⛔ ${e.blocked}</span>`
            : routes.length ? html`<span class="muted">add another ${e.kind === 'http' ? 'hostname' : 'host port'}</span>`
              : html`<span class="muted">unbound — not reachable</span>`;
          return html`
            ${routes.map((r, i) => html`<tr class="ing-route" data-ep=${e.component + '.' + e.slot}>
              ${head(i)}
              <td class="mono">${r.source}</td>
              <td class="mono">${routeText(e, r)}</td>
              <td><button class="act rm" title="remove this route" @click=${() => this._ingRemove(e, r)}>remove</button></td>
              <td>${e.blocked ? html`<span class="st-failed">⛔ ${e.blocked}</span>` : routeState(e, r)}</td></tr>`)}
            <tr class="ing-add" data-ep=${e.component + '.' + e.slot}>
              ${head(routes.length)}
              <td><select ?disabled=${!!e.blocked} @change=${(ev) => this._ingSetEdit(e, { source: ev.target.value })}>
                <option value="" ?selected=${!ed.source}>${routes.length ? '+ route via…' : '— unbound —'}</option>
                ${sources(e.kind).map((s) => html`<option value=${s} ?selected=${ed.source === s}>${s}</option>`)}
              </select></td>
              <td>${routeEd}</td>
              <td>${ed.source ? html`<button class="act go" @click=${() => this._ingPublish(e)}>${routes.length ? 'add' : 'publish'}</button>` : nothing}</td>
              <td>${addState}</td></tr>`;
        })}
        ${exposes.length === 0 ? html`<tr><td class="muted" colspan="6">no matching endpoints</td></tr>` : nothing}
      </table>`}`;
  }

  // ---- ingress → endpoints: live routes, listeners, terminators ----
  _ingressEndpointsView() {
    const d = this._ingress;
    if (!d) return html`<span class="muted">loading…</span>`;
    const routes = (d.routes || []).filter((r) => this._match(r.host, r.component, r.slot, r.source));
    const streams = d.streams || [];
    const forwards = d.forwards || [];
    const lst = d.httpListener || {};
    return html`
      <p class="muted">The live public routing table — what the outside can reach right now. Publish
        or unpublish under <a class="link" @click=${() => this._emit('bx-admin-tab', 'expose')}>services / expose</a>.</p>
      <div class="muted" style="margin:2px 0 10px">
        builtin HTTP listener: ${lst.listen ? html`<b>${lst.listen}</b> (${lst.tls ? 'TLS' : 'no TLS — front it, or use the Traefik tile'})` : 'off — start xbind with --ingress-listen'}
      </div>
      <h4>HTTP routes</h4>
      ${this._filterBar('filter routes by host or tile…', null, routes.length, (d.routes || []).length)}
      <table class="tbl">
        <tr><th>public host</th><th>→ tile</th><th>via</th></tr>
        ${routes.map((r) => html`<tr>
          <td class="mono">${r.host}</td>
          <td class="mono">${r.component}.${r.slot}</td>
          <td>${r.source}${r.zone ? html` <span class="muted">(zone ${r.zone})</span>` : nothing}</td></tr>`)}
        ${routes.length === 0 ? html`<tr><td class="muted" colspan="3">no HTTP routes${this._q ? ' match' : ''}</td></tr>` : nothing}
      </table>
      ${streams.length ? html`<h4>stream listeners (tcp / udp)</h4>
        <table class="tbl">
          <tr><th>host listen</th><th>→ tile</th><th>proto</th><th>state</th></tr>
          ${streams.map((s) => html`<tr>
            <td class="mono">${s.listen}</td>
            <td class="mono">${s.component}.${s.slot} → :${s.port}</td>
            <td>${s.proto}</td>
            <td>${s.error ? html`<span class="st-failed">⚠ ${s.error}</span>` : html`<span class="st-healthy">${s.active} active</span>`}</td></tr>`)}
        </table>` : nothing}
      ${forwards.length ? html`<h4>terminator forward doors</h4>
        <table class="tbl">
          <tr><th>terminator tile</th><th>state</th></tr>
          ${forwards.map((f) => html`<tr>
            <td class="mono">${f.source}</td>
            <td>${f.error ? html`<span class="st-failed">⚠ ${f.error}</span>` : html`<span class="st-healthy">up</span>`}</td></tr>`)}
        </table>` : nothing}`;
  }
}

customElements.define('bx-admin-ingress', BxAdminIngress);
